src/write.jl

```diff
@@ -2,28 +2,30 @@
 ##### utilities
 #####
 
-_edf_write(io::IO, value::Union{String,Char}) = Base.write(io, value)
-_edf_write(io::IO, date::Date) = Base.write(io, uppercase(Dates.format(date, dateformat"dd-u-yyyy")))
-_edf_write(io::IO, date::DateTime) = Base.write(io, Dates.format(date, dateformat"dd\.mm\.yyHH\.MM\.SS"))
-_edf_write(io::IO, x::Integer) = Base.write(io, string(x))
-_edf_write(io::IO, x::AbstractFloat) = Base.write(io, string(isinteger(x) ? trunc(Int, x) : x))
+_edf_repr(value::Union{String,Char}) = value
+_edf_repr(date::Date) = uppercase(Dates.format(date, dateformat"dd-u-yyyy"))
+_edf_repr(date::DateTime) = Dates.format(date, dateformat"dd\.mm\.yyHH\.MM\.SS")
+_edf_repr(x::Integer) = string(x)
+_edf_repr(x::AbstractFloat) = string(isinteger(x) ? trunc(Int, x) : x)
+_edf_repr(::Missing) = 'X'
 
-function _edf_write(io::IO, metadata::T) where T<:Union{PatientID,RecordingID}
-    bytes_written = 0
-    for name in fieldnames(T)
-        field = getfield(metadata, name)
-        if T <: RecordingID && name === :startdate
-            bytes_written += _edf_write(io, "Startdate ")
-        end
-        bytes_written += field isa Missing ? Base.write(io, 'X') : _edf_write(io, field)
-        bytes_written += Base.write(io, ' ')
-    end
-    return bytes_written
+function _edf_repr(metadata::T) where T<:Union{PatientID,RecordingID}
+    header = T <: RecordingID ? String["Startdate"] : String[]
+    return join([header; [_edf_repr(getfield(metadata, name)) for name in fieldnames(T)]], ' ')
 end
 
-function edf_write(io::IO, value, byte_limit::Integer)
-    bytes_written = _edf_write(io, value)
-    bytes_written <= byte_limit || error("Written value $value contains more bytes than limit $byte_limit")
+function edf_write(io::IO, value, byte_limit::Integer; truncate::Bool=true)
+    edf_value = _edf_repr(value)
+    @assert isascii(string(edf_value))
+    size = length(edf_value)
+    if size > byte_limit
+        if truncate
+            edf_value = chop(edf_value; head=0, tail=(size - byte_limit))
+        else
+            error("$value is $(sizeof(edf_value)) bytes. The byte limit for this value is $byte_limit")
```

Comment:
Note that you're using `sizeof` here but `length` for the check. They're equivalent for ASCII, and indeed ASCII is required per the EDF spec, just thought I'd point it out.